com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/CreateOrUpdateAppStep.java

```diff
@@ -0,0 +1,419 @@
+import org.springframework.stereotype.Component;
+
+import com.sap.cloud.lm.sl.cf.client.lib.domain.CloudApplicationExtended;
+import com.sap.cloud.lm.sl.cf.client.lib.domain.CloudApplicationExtended.AttributeUpdateStrategy;
+import com.sap.cloud.lm.sl.cf.client.lib.domain.CloudServiceExtended;
+import com.sap.cloud.lm.sl.cf.client.lib.domain.ServiceKeyToInject;
+import com.sap.cloud.lm.sl.cf.core.helpers.MtaArchiveElements;
+import com.sap.cloud.lm.sl.cf.core.security.serialization.SecureSerializationFacade;
+import com.sap.cloud.lm.sl.cf.core.util.ApplicationConfiguration;
+import com.sap.cloud.lm.sl.cf.persistence.processors.DefaultFileDownloadProcessor;
+import com.sap.cloud.lm.sl.cf.persistence.services.FileContentProcessor;
+import com.sap.cloud.lm.sl.cf.persistence.services.FileStorageException;
+import com.sap.cloud.lm.sl.cf.process.Constants;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+import com.sap.cloud.lm.sl.cf.process.util.ApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.ApplicationAttributeUpdater.UpdateState;
+import com.sap.cloud.lm.sl.cf.process.util.DiskQuotaApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.ElementUpdater.UpdateBehavior;
+import com.sap.cloud.lm.sl.cf.process.util.EnvironmentApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.MemoryApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.ServiceOperationUtil;
+import com.sap.cloud.lm.sl.cf.process.util.StagingApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.cf.process.util.UrisApplicationAttributeUpdater;
+import com.sap.cloud.lm.sl.common.SLException;
+import com.sap.cloud.lm.sl.common.util.JsonUtil;
+import com.sap.cloud.lm.sl.common.util.MapUtil;
+import com.sap.cloud.lm.sl.mta.handlers.ArchiveHandler;
+import com.sap.cloud.lm.sl.mta.util.ValidatorUtil;
+
+@Component("createOrUpdateAppStep")
+@Scope(BeanDefinition.SCOPE_PROTOTYPE)
+public class CreateOrUpdateAppStep extends SyncFlowableStep {
+
+    private SecureSerializationFacade secureSerializer = new SecureSerializationFacade();
+
+    @Autowired
+    protected ApplicationConfiguration configuration;
+
+    @Override
+    protected StepPhase executeStep(ExecutionWrapper execution) throws FileStorageException {
+        // Get the next cloud application from the context:
+        CloudApplicationExtended app = StepsUtil.getApp(execution.getContext());
+        try {
+            // Check if an application with this name already exists
+            CloudControllerClient client = execution.getControllerClient();
+            CloudApplication existingApp = client.getApplication(app.getName(), false);
+            StepsUtil.setExistingApp(execution.getContext(), existingApp);
+
+            boolean servicesPropertiesChanged = true;
+            boolean appPropertiesChanged = true;
+            boolean userPropertiesChanged = true;
+
+            printStepStartMessage(app, existingApp == null);

```

Comment:
move the result of existingApp==null check to a, instance variable, since it seems we need this check a lot